contracts/core/Registry.sol

```diff
@@ -1,47 +1,174 @@
+        bytes32 memberRole = _generateRole(_identityId, RoleType.MEMBER);
 
-    // update the metadata of the identity
-    // checks ownership role first, probably separate roles for this vs using it (owner vs user?)
-    function updateIdentityMetadata(uint _identityId, string memory _metadata) external {
+        if (
+            !hasRole(ownerRole, msg.sender) ||
+            !hasRole(memberRole, msg.sender)
+        ) {
+            revert NO_ACCESS_TO_ROLE();
+        }
 
+        identities[_identityId].metadata = _metadata;
+
+        emit IdentityMetadataUpdated(_identityId, _metadata);
     }
 
-    // this will use solmate ROLES to check if the msg.sender is an owner of the project
-    // @todo figure out how to best represent project ownership and details
-    function isOwnerOfProject(uint _identityId, address _owner) external view returns (bool) {
-        
+    /// @notice Returns if the given address is an owner of the identity
+    function isOwnerOfIdentity(uint _identityId, address _owner) external view returns (bool) {
+        bytes32 ownerRole = _generateRole(_identityId, RoleType.OWNER);
+        return hasRole(ownerRole, _owner);
+    }
+
+
+    /// @notice Returns if the given address is an member of the identity
+    function isMemberOfIdentity(uint _identityId, address _owner) external view returns (bool) {
+        bytes32 memberRole = _generateRole(_identityId, RoleType.MEMBER);
+        return hasRole(memberRole, _owner);
+    }
+
+    /// @notice Generates the attestation address for the given identity
+    function _generateAttestationAddr(uint _identityId, string memory _name) internal pure returns (address) {
+        bytes32 attestationHash = keccak256(
+            abi.encodePacked(_identityId, _name)
+        );
+
+        return address(uint160(uint256(attestationHash)));
+    }
+
+    /// @notice Generates the OZ role for an given identity
+    function _generateRole(uint _identityId, RoleType _roleType) internal pure returns (bytes32 roleHash) {
+        roleHash = keccak256(
+            abi.encodePacked(_identityId, _roleType)
+        );
+    }
+
+    // --- ACCESS CONTROL ---
+
+    /// @notice OZ function to grant role reverts 
+    function grantRole(bytes32 role, address account) public virtual override onlyRole(getRoleAdmin(role)) {
+        revert NOT_SUPPORTED();
+    }
+
+    /// @notice OZ function to grant role reverts 
+    function revokeRole(bytes32 role, address account) public virtual override onlyRole(getRoleAdmin(role)) {
+        revert NOT_SUPPORTED();
+    }
+
+    function grantIdentityRole(uint _identityId, RoleType _roleType, address account) external returns (uint) {
+        bytes32 role = _generateRole(_identityId, _roleType);
+        _grantRole(role, account);
+
+        if(_roleType == RoleType.OWNER) {
+            IdentityDetails memory identity = identities[_identityId];
+            // TODO: HOW to update attestation address cause name and identityId is never updated?
```

Comment:
@zobront @codenamejason  @KurtMerbeth 
I think we would _generateAttestationAddr to take an extra parameter which relies on an address or something.

Reply:
https://github.com/allo-protocol/allo-v2/pull/12#discussion_r1238885697